### PR TITLE
followup [add SparkSQLEngine idle check]

### PR DESCRIPTION
**Pull request by xjzrc**
修复SparkSQLEngine停止时没有关闭timeoutChecker线程池，以及增加timeoutChecker打印日志

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 1]
```diff
@@ -41,21 +41,32 @@ private[spark] final class SparkSQLEngine(name: String, spark: SparkSession)
...
     val checkTask = new Runnable {
       override def run(): Unit = {
-        val current = System.currentTimeMillis
         val sessionManager = backendService.sessionManager.asInstanceOf[SparkSQLSessionManager]
-        if (sessionManager.getOpenSessionCount <= 0 &&
-          current - sessionManager.getLogoutTime > idleTimeout) {
+        val openSessionCount = sessionManager.getOpenSessionCount
+        val logoutTime = sessionManager.getLogoutTime
+        val current = System.currentTimeMillis
+        if (logger.isDebugEnabled) {
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:59]: this line is redundent

Comment (xjzrc) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala]: 已修改，还有关注下[打包时jars目录重复](https://github.com/yaooqinn/kyuubi/pull/251/commits/29e9dd45e90b1549b68e96024b1ee866429bf40d)

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 2]
```diff
@@ -41,21 +41,30 @@ private[spark] final class SparkSQLEngine(name: String, spark: SparkSession)
...
-        val current = System.currentTimeMillis
         val sessionManager = backendService.sessionManager.asInstanceOf[SparkSQLSessionManager]
-        if (sessionManager.getOpenSessionCount <= 0 &&
-          current - sessionManager.getLogoutTime > idleTimeout) {
+        val openSessionCount = sessionManager.getOpenSessionCount
+        val logoutTime = sessionManager.getLogoutTime
+        val current = System.currentTimeMillis
+        debug(s"timeoutChecker, openSessionCount:$openSessionCount," +
+          s" current time:$current, logoutTime:$logoutTime, idleTimeout:$idleTimeout")
+        if (openSessionCount <= 0 && (current - logoutTime) > idleTimeout) {
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:61]: Not only related to this PR but also the original one, there is a concurrent issue here. When we are closing a session, the `openSessionCount` could be set to 0, but the `logoutTime` here hasn't been updated, then the whole engine will self terminated unexpectedly.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 3]
```diff
@@ -41,21 +41,30 @@ private[spark] final class SparkSQLEngine(name: String, spark: SparkSession)
...
+          info(s"Stopping idle $name[${spark.conf.get("spark.app.name")}]")
           sys.exit(0)
         }
       }
     }
+    info(s"Scheduling $name cleaning every ${interval / 1000} seconds")
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:67]: this is not very necessary and maybe a little bit annoying

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 4]
```diff
@@ -41,21 +41,30 @@ private[spark] final class SparkSQLEngine(name: String, spark: SparkSession)
...
         val sessionManager = backendService.sessionManager.asInstanceOf[SparkSQLSessionManager]
-        if (sessionManager.getOpenSessionCount <= 0 &&
-          current - sessionManager.getLogoutTime > idleTimeout) {
+        val openSessionCount = sessionManager.getOpenSessionCount
+        val logoutTime = sessionManager.getLogoutTime
+        val current = System.currentTimeMillis
+        debug(s"timeoutChecker, openSessionCount:$openSessionCount," +
+          s" current time:$current, logoutTime:$logoutTime, idleTimeout:$idleTimeout")
+        if (openSessionCount <= 0 && (current - logoutTime) > idleTimeout) {
+          info(s"Stopping idle $name[${spark.conf.get("spark.app.name")}]")
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:62]: This log is inside the engine,  let's make the note as simple as possible
e.g. `info(s"Idled for more than $idleTimeout, terminating")`

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 2 above)
Comment (xjzrc) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:61]: If we update the `logoutTime` first, then remove the `Session`?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 3 above)
Comment (xjzrc) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:67]: ok, I will remove it.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 4 above)
Comment (xjzrc) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:62]: ok, I will fix it.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 2 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:61]: make sense

**Comment by yaooqinn** [comment]
thanks, merging to master.

**Pull request closed without merging**